x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java

```diff
@@ -338,4 +333,39 @@ public static void checkForCcsLicense(
             }
         }
     }
+
+    /**
+     * Mark cluster with a default cluster state with the given status and potentially failure from exception.
+     * Most metrics are set to 0 except for "took" which is set to the total time taken so far.
+     * The status must be the final state of the cluster, not RUNNING.
+     */
+    public static void markClusterWithFinalStateAndNoShards(
+        EsqlExecutionInfo executionInfo,
+        String clusterAlias,
+        Cluster.Status status,
+        @Nullable Exception ex
+    ) {
+        assert status != Cluster.Status.RUNNING : "status must be a final state, not RUNNING";
+        executionInfo.swapCluster(clusterAlias, (k, v) -> {
+            Cluster.Builder builder = new Cluster.Builder(v).setStatus(status)
+                .setTook(executionInfo.tookSoFar())
+                .setTotalShards(Objects.requireNonNullElse(v.getTotalShards(), 0))
+                .setSuccessfulShards(Objects.requireNonNullElse(v.getTotalShards(), 0))
+                .setSkippedShards(Objects.requireNonNullElse(v.getTotalShards(), 0))
+                .setFailedShards(Objects.requireNonNullElse(v.getTotalShards(), 0));
+            if (ex != null) {
+                builder.setFailures(List.of(new ShardSearchFailure(ex)));
+            }
+            return builder.build();
+        });
+    }
+
+    public static boolean shouldIgnoreRuntimeError(EsqlExecutionInfo executionInfo, String clusterAlias, Exception e) {
+        if (executionInfo.isSkipUnavailable(clusterAlias) == false) {
+            return false;
+        }
+
+        return ExceptionsHelper.isRemoteUnavailableException(e)
+            || (e instanceof RemoteTransportException && e.getCause() instanceof TaskCancelledException);
```

Comment:
Why do we include TaskCancelledException here?

Reply:
I was thinking we didn't fail when remote is cancelled? If it's the case we need it.